mobile/v1/bigquery/sql-runner/sql/standard/00-setup/01-main/01-functions-and-procedures.sql

```diff
@@ -0,0 +1,305 @@
+
+      FROM column_field_paths
+      GROUP BY 1
+    )
+
+    , column_coalesce_result AS (
+    SELECT 
+      ARRAY_TO_STRING(
+      ARRAY_AGG(CONCAT('COALESCE(', ARRAY_TO_STRING(paths, ', '), ', NULL) AS ',  fm.target_field)), ', ') AS column_coalesce
+
+    FROM fields f
+    LEFT JOIN field_mapping fm
+    ON f.field_name = fm.source_field
+    )
+
+    -- not used presently but leaving in for future
+    -- , column_versions_result AS (
+    -- SELECT
+    --   ARRAY_TO_STRING(ARRAY_AGG(DISTINCT column_name),',') AS column_versions
+
+    -- FROM column_field_paths
+    -- )
+
+    SELECT
+      column_coalesce
+
+    FROM column_coalesce_result
+
+    """);
+
+  EXECUTE IMMEDIATE COLUMN_VERSIONS_QUERY INTO COLUMN_COALESCE;
+
+END;
+
+-- Used to dynamically select the correct mobile context fields from events table.
+-- MOBILE_CONTEXT_COLUMNS = Coalesce of fields within context e.g. "COALESCE(contexts_com_snowplowanalytics_snowplow_mobile_context_1_0_1[SAFE_OFFSET(0)].device_manufacturer, contexts_com_snowplowanalytics_snowplow_mobile_context_1_0_0[SAFE_OFFSET(0)].device_manufacturer) AS device_manufacturer"
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.mobile_mobile_context_fields (context_enabled BOOLEAN,
+                                                                             OUT MOBILE_CONTEXT_COLUMNS STRING)
+
+BEGIN
+  --If mobile context enabled, find all fields across context schema versions, else return NULL fields.
+  IF context_enabled THEN
+
+    CALL {{.output_schema}}.combine_column_versions(
+                    '{{.input_schema}}',     -- source_schema
+                    'events',                -- source_table
+                    '["device_manufacturer","device_model","os_type","os_version","android_idfa","apple_idfa","apple_idfv","carrier","open_idfa","network_technology","network_type"]', --source_fields. Quoted array to allow for concat
+                    'contexts_com_snowplowanalytics_snowplow_mobile_context_1_0', -- columns_prefix
+                    false,                   -- rename_y_n
+                    '[""]',                  -- rename_fields
+                    MOBILE_CONTEXT_COLUMNS   -- returned context coalesce columns
+                    );
+
+  ELSE
+
+    SET MOBILE_CONTEXT_COLUMNS = (
+      """CAST(NULL AS STRING) AS device_manufacturer,
+      CAST(NULL AS STRING) AS device_model,
+      CAST(NULL AS STRING) AS os_type,
+      CAST(NULL AS STRING) AS os_version,
+      CAST(NULL AS STRING) AS android_idfa,
+      CAST(NULL AS STRING) AS apple_idfa,
+      CAST(NULL AS STRING) AS apple_idfv,
+      CAST(NULL AS STRING) AS carrier,
+      CAST(NULL AS STRING) AS open_idfa,
+      CAST(NULL AS STRING) AS network_technology,
+      CAST(NULL AS STRING) AS network_type""");
+
+  END IF;
+
+END;
+
+
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.mobile_app_errors_fields (OUT APP_ERRORS_EVENTS_COLUMNS STRING)
```

Comment:
As we mentioned on the call, it seems odd to have procedures that simply wrap other procedures. Worth considering whether there's a better way (perhaps we call combine_column_versions into a variable at the beginning of the step in which we use it, for example)